ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/UserManagementServiceIT.scala

```diff
@@ -109,71 +122,199 @@ final class UserManagementServiceIT extends LedgerTestSuite {
-      createAgainError <- ledger.userManagement
-        .createUser(CreateUserRequest(Some(User("a", "b")), Nil))
-        .mustFail("allocating a duplicate user")
-
-      getUserResult <- ledger.userManagement.getUser(GetUserRequest("a"))
-      getUserError <- ledger.userManagement
-        .getUser(GetUserRequest("b"))
-        .mustFail("retrieving non-existent user")
-
-      grantResult <- ledger.userManagement.grantUserRights(
-        GrantUserRightsRequest(
-          "a",
-          List(Permission(Permission.Kind.ParticipantAdmin(Permission.ParticipantAdmin()))),
-        )
-      )
-      listRightsResult <- ledger.userManagement.listUserRights(ListUserRightsRequest("a"))
-      revokeResult <- ledger.userManagement.revokeUserRights(
-        RevokeUserRightsRequest(
-          "a",
-          List(Permission(Permission.Kind.ParticipantAdmin(Permission.ParticipantAdmin()))),
-        )
-      )
-      _ <- ledger.userManagement.deleteUser(DeleteUserRequest("a"))
-    } yield {
+    val adminPermission =
+      Permission(Permission.Kind.ParticipantAdmin(Permission.ParticipantAdmin()))
+    val actAsPermission1 =
+      Permission(Permission.Kind.CanActAs(Permission.CanActAs("acting-party-1")))
+    val readAsPermission1 =
+      Permission(Permission.Kind.CanReadAs(Permission.CanReadAs("reading-party-1")))
+
+    val userRightsBatch = List(
+      actAsPermission1,
+      Permission(Permission.Kind.CanActAs(Permission.CanActAs("acting-party-2"))),
+      readAsPermission1,
+      Permission(Permission.Kind.CanReadAs(Permission.CanReadAs("reading-party-2"))),
+    )
+
+    def assertUserNotFound(t: Throwable): Unit = {
       assertGrpcError(
-        ledger,
-        createAgainError,
-        Status.Code.ALREADY_EXISTS,
-        LedgerApiErrors.AdminServices.UserAlreadyExists,
-        None,
+        participant = ledger,
+        t = t,
+        expectedCode = Status.Code.NOT_FOUND,
+        selfServiceErrorCode = LedgerApiErrors.AdminServices.UserNotFound,
+        exceptionMessageSubstring = None,
       )
+    }
+
+    def assertUserAlreadyExists(t: Throwable): Unit = {
       assertGrpcError(
-        ledger,
-        getUserError,
-        Status.Code.NOT_FOUND,
-        LedgerApiErrors.AdminServices.UserNotFound,
-        None,
+        participant = ledger,
+        t = t,
+        expectedCode = Status.Code.ALREADY_EXISTS,
+        selfServiceErrorCode = LedgerApiErrors.AdminServices.UserAlreadyExists,
+        exceptionMessageSubstring = None,
       )
-      assert(createResult == User("a", "b"))
-      assert(getUserResult == User("a", "b"))
-      assert(
-        grantResult.newlyGrantedRights == List(
-          Permission(Permission.Kind.ParticipantAdmin(Permission.ParticipantAdmin()))
+    }
+
+    def testCreateUser(): Future[Unit] = {
+      for {
+        res1 <- ledger.userManagement.createUser(
+          CreateUserRequest(Some(User("user1", "party1")), Nil)
```

Comment:
Could you extract the different User records into vals. There is a lot of repetition ripe for future mistakes.

Reply:
Also, as mentioned above
There is a prior art of how we generate such strings that guarantee isolation and at the same time allow us to debug logfiles/database contents when something goes wrong. Please have a look at this and adopt the code:
https://github.com/digital-asset/daml/blob/d4ebce6f4446c9f4744ed610cfb83161a8640d16/ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala#L134

Reply:
Ok